# CICIDS

### **random_forest.ipynb**

The CICIDS2017 dataset is a comprehensive dataset for network intrusion detection, created by the Canadian Institute for Cybersecurity. It includes a diverse set of attack scenarios and normal traffic, making it suitable for training and evaluating intrusion detection systems.

The dataset includes various types of attacks such as Brute Force, Heartbleed, Botnet, DoS (Denial of Service), DDoS (Distributed Denial of Service), Web attacks, and Infiltration of the network from inside.

In [27]:
model_name = "random_forest"

In [1]:
import warnings
from sklearn.exceptions import UndefinedMetricWarning
warnings.filterwarnings("ignore", category=UndefinedMetricWarning)

## Step 1. Read data and import necessary libraries

In [2]:
import pandas as pd
df_train = pd.read_csv("./data/concatenated/concat.csv")

In [3]:
df_train.head(5)

Destination Port   Flow Duration   Total Fwd Packets  \
0              49188               4                   2   
1              49188               1                   2   
2              49188               1                   2   
3              49188               1                   2   
4              49486               3                   2   

    Total Backward Packets  Total Length of Fwd Packets  \
0                        0                           12   
1                        0                           12   
2                        0                           12   
3                        0                           12   
4                        0                           12   

    Total Length of Bwd Packets   Fwd Packet Length Max  \
0                             0                       6   
1                             0                       6   
2                             0                       6   
3                             0                       6   
4                             0                       6   

    Fwd Packet Length Min   Fwd Packet Length Mean   Fwd Packet Length Std  \
0                       6                      6.0                     0.0   
1                       6                      6.0                     0.0   
2                       6                      6.0                     0.0   
3                       6                      6.0                     0.0   
4                       6                      6.0                     0.0   

   ...   min_seg_size_forward  Active Mean   Active Std   Active Max  \
0  ...                     20          0.0          0.0            0   
1  ...                     20          0.0          0.0            0   
2  ...                     20          0.0          0.0            0   
3  ...                     20          0.0          0.0            0   
4  ...                     20          0.0          0.0            0   

    Active Min  Idle Mean   Idle Std   Idle Max   Idle Min   Label  
0            0        0.0        0.0          0          0  BENIGN  
1            0        0.0        0.0          0          0  BENIGN  
2            0        0.0        0.0          0          0  BENIGN  
3            0        0.0        0.0          0          0  BENIGN  
4            0        0.0        0.0          0          0  BENIGN  

[5 rows x 79 columns]

In [4]:
df_train.shape

(2830743, 79)

In [5]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2830743 entries, 0 to 2830742
Data columns (total 79 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0    Destination Port             int64  
 1    Flow Duration                int64  
 2    Total Fwd Packets            int64  
 3    Total Backward Packets       int64  
 4   Total Length of Fwd Packets   int64  
 5    Total Length of Bwd Packets  int64  
 6    Fwd Packet Length Max        int64  
 7    Fwd Packet Length Min        int64  
 8    Fwd Packet Length Mean       float64
 9    Fwd Packet Length Std        float64
 10  Bwd Packet Length Max         int64  
 11   Bwd Packet Length Min        int64  
 12   Bwd Packet Length Mean       float64
 13   Bwd Packet Length Std        float64
 14  Flow Bytes/s                  float64
 15   Flow Packets/s               float64
 16   Flow IAT Mean                float64
 17   Flow IAT Std                 float64
 18   Flow IAT Max         

In [6]:
df_train.describe()

c:\Users\Harman\AppData\Local\Programs\Python\Python312\Lib\site-packages\pandas\core\nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)
c:\Users\Harman\AppData\Local\Programs\Python\Python312\Lib\site-packages\pandas\core\nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)


Destination Port   Flow Duration   Total Fwd Packets  \
count       2.830743e+06    2.830743e+06        2.830743e+06   
mean        8.071483e+03    1.478566e+07        9.361160e+00   
std         1.828363e+04    3.365374e+07        7.496728e+02   
min         0.000000e+00   -1.300000e+01        1.000000e+00   
25%         5.300000e+01    1.550000e+02        2.000000e+00   
50%         8.000000e+01    3.131600e+04        2.000000e+00   
75%         4.430000e+02    3.204828e+06        5.000000e+00   
max         6.553500e+04    1.200000e+08        2.197590e+05   

        Total Backward Packets  Total Length of Fwd Packets  \
count             2.830743e+06                 2.830743e+06   
mean              1.039377e+01                 5.493024e+02   
std               9.973883e+02                 9.993589e+03   
min               0.000000e+00                 0.000000e+00   
25%               1.000000e+00                 1.200000e+01   
50%               2.000000e+00                 6.200000e+01   
75%               4.000000e+00                 1.870000e+02   
max               2.919220e+05                 1.290000e+07   

        Total Length of Bwd Packets   Fwd Packet Length Max  \
count                  2.830743e+06            2.830743e+06   
mean                   1.616264e+04            2.075999e+02   
std                    2.263088e+06            7.171848e+02   
min                    0.000000e+00            0.000000e+00   
25%                    0.000000e+00            6.000000e+00   
50%                    1.230000e+02            3.700000e+01   
75%                    4.820000e+02            8.100000e+01   
max                    6.554530e+08            2.482000e+04   

        Fwd Packet Length Min   Fwd Packet Length Mean  \
count            2.830743e+06             2.830743e+06   
mean             1.871366e+01             5.820194e+01   
std              6.033935e+01             1.860912e+02   
min              0.000000e+00             0.000000e+00   
25%              0.000000e+00             6.000000e+00   
50%              2.000000e+00             3.400000e+01   
75%              3.600000e+01             5.000000e+01   
max              2.325000e+03             5.940857e+03   

        Fwd Packet Length Std  ...   act_data_pkt_fwd   min_seg_size_forward  \
count            2.830743e+06  ...       2.830743e+06           2.830743e+06   
mean             6.891013e+01  ...       5.418218e+00          -2.741688e+03   
std              2.811871e+02  ...       6.364257e+02           1.084989e+06   
min              0.000000e+00  ...       0.000000e+00          -5.368707e+08   
25%              0.000000e+00  ...       0.000000e+00           2.000000e+01   
50%              0.000000e+00  ...       1.000000e+00           2.400000e+01   
75%              2.616295e+01  ...       2.000000e+00           3.200000e+01   
max              7.125597e+03  ...       2.135570e+05           1.380000e+02   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  2.830743e+06  2.830743e+06  2.830743e+06  2.830743e+06  2.830743e+06   
mean   8.155132e+04  4.113412e+04  1.531825e+05  5.829582e+04  8.316037e+06   
std    6.485999e+05  3.933815e+05  1.025825e+06  5.770923e+05  2.363008e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.100000e+08  7.420000e+07  1.100000e+08  1.100000e+08  1.200000e+08   

           Idle Std      Idle Max      Idle Min  
count  2.830743e+06  2.830743e+06  2.830743e+06  
mean   5.038439e+05  8.695752e+06  7.920031e+06  
std    4.602984e+06  2.436689e+07  2.336342e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.00

In [7]:
# Remove leading and trailing whitespaces from column names
df_train.columns = df_train.columns.str.strip()

In [8]:
df_train.columns

Index(['Destination Port', 'Flow Duration', 'Total Fwd Packets',
       'Total Backward Packets', 'Total Length of Fwd Packets',
       'Total Length of Bwd Packets', 'Fwd Packet Length Max',
       'Fwd Packet Length Min', 'Fwd Packet Length Mean',
       'Fwd Packet Length Std', 'Bwd Packet Length Max',
       'Bwd Packet Length Min', 'Bwd Packet Length Mean',
       'Bwd Packet Length Std', 'Flow Bytes/s', 'Flow Packets/s',
       'Flow IAT Mean', 'Flow IAT Std', 'Flow IAT Max', 'Flow IAT Min',
       'Fwd IAT Total', 'Fwd IAT Mean', 'Fwd IAT Std', 'Fwd IAT Max',
       'Fwd IAT Min', 'Bwd IAT Total', 'Bwd IAT Mean', 'Bwd IAT Std',
       'Bwd IAT Max', 'Bwd IAT Min', 'Fwd PSH Flags', 'Bwd PSH Flags',
       'Fwd URG Flags', 'Bwd URG Flags', 'Fwd Header Length',
       'Bwd Header Length', 'Fwd Packets/s', 'Bwd Packets/s',
       'Min Packet Length', 'Max Packet Length', 'Packet Length Mean',
       'Packet Length Std', 'Packet Length Variance', 'FIN Flag Count',
       'SYN Flag Co

## Step 2. Data Cleaning

### A. Missing values

In [9]:
print(df_train.isna().sum().sum())

1358


In [10]:
df_train.dropna(subset=["Flow Bytes/s"], inplace=True)

In [11]:
print(df_train.isna().sum().sum())

0


### Inf. values

In [12]:
import numpy as np
df_train = df_train.replace([np.inf, -np.inf], np.nan).dropna()

## Step 3. Data Preparation

### A. Normalise numeric features

In [13]:
# Get all numerical columns
numerical_columns = df_train.select_dtypes(include="number").columns

In [14]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df_train[numerical_columns] = scaler.fit_transform(df_train[numerical_columns])

### B. Map Labels to Multi-class

In [15]:
df_train["Label"].value_counts()

Label
BENIGN                        2271320
DoS Hulk                       230124
PortScan                       158804
DDoS                           128025
DoS GoldenEye                   10293
FTP-Patator                      7935
SSH-Patator                      5897
DoS slowloris                    5796
DoS Slowhttptest                 5499
Bot                              1956
Web Attack � Brute Force         1507
Web Attack � XSS                  652
Infiltration                       36
Web Attack � Sql Injection         21
Heartbleed                         11
Name: count, dtype: int64

In [16]:
attack_mapping = {
	"BENIGN": 0,
	"DoS Hulk": 1,
	"PortScan": 2,
	"DDoS": 3,
	"DoS GoldenEye": 4,
	"FTP-Patator": 5,
	"SSH-Patator": 6,
	"DoS slowloris": 7,
	"DoS Slowhttptest": 8,
	"Bot": 9,
	"Web Attack � Brute Force": 10,
	"Web Attack � XSS": 11,
	"Infiltration": 12,
	"Web Attack � Sql Injection": 13,
	"Heartbleed": 14,
}

df_train["Label"] = df_train["Label"].map(attack_mapping)

In [17]:
df_train["Label"].value_counts()

Label
0     2271320
1      230124
2      158804
3      128025
4       10293
5        7935
6        5897
7        5796
8        5499
9        1956
10       1507
11        652
12         36
13         21
14         11
Name: count, dtype: int64

### C. Data Splitting

In [18]:
X = df_train.drop(columns="Label")
y = df_train["Label"]

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### D. Apply SMOTE to balance the training data

In [20]:
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler

# 1. Undersample the majority class
undersampling_strategy = {
    0: 1000,
    1: 1000,
    2: 1000,
    3: 1000,
    4: 1000,
    5: 1000,
    6: 1000,
    7: 1000,
    8: 1000,
    9: 1000,
	10: 1000,
}
rus = RandomUnderSampler(random_state=42, sampling_strategy=undersampling_strategy)
X_train_undersampled, y_train_undersampled = rus.fit_resample(X_train, y_train)

# 2. Oversample the minority class
smote = SMOTE(random_state=42, sampling_strategy="auto")
X_train_balanced, y_train_balanced = smote.fit_resample(X_train_undersampled, y_train_undersampled)

In [21]:
# Check class distribution after SMOTE
from collections import Counter

print(f"Class distribution before SMOTE: {Counter(y_train)}")
print(f"Class distribution after SMOTE: {Counter(y_train_balanced)}")

Class distribution before SMOTE: Counter({0: 1817112, 1: 184342, 2: 126927, 3: 102239, 4: 8219, 5: 6363, 6: 4769, 7: 4630, 8: 4390, 9: 1515, 10: 1206, 11: 533, 12: 29, 13: 17, 14: 9})
Class distribution after SMOTE: Counter({0: 1000, 1: 1000, 2: 1000, 3: 1000, 4: 1000, 5: 1000, 6: 1000, 7: 1000, 8: 1000, 9: 1000, 10: 1000, 11: 1000, 12: 1000, 13: 1000, 14: 1000})


## Step 4. Model

#### A. Defining the model

In [22]:
from sklearn.metrics import classification_report, accuracy_score, f1_score
from sklearn.model_selection import RandomizedSearchCV, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from tqdm import tqdm
import numpy as np

In [ ]:
param_dist = {
	"n_estimators": np.arange(260, 321, 10),
	"max_depth": [37, 38, 39],
	"min_samples_split": [5],
	"min_samples_leaf": [1],
	"max_features": ["sqrt"],
	"bootstrap": [True],
}

# Best parameters found:  {
    # 'n_estimators': 280, 
    # 'min_samples_split': 5, 
    # 'min_samples_leaf': 1, 
    # 'max_features': 'sqrt', 
    # 'max_depth': 38, 
    # 'bootstrap': True
# }

rf = RandomForestClassifier(random_state=42)

random_search = RandomizedSearchCV(
	rf, param_distributions=param_dist,
	n_iter=250, cv=5,
	n_jobs=-1, verbose=2,
	random_state=42, return_train_score=True)

random_search.fit(X_train_balanced, y_train_balanced)

# 4000 fits -> 31m 14s
# 1250 fits -> 14m 18s		12m 59s
#  105 fits ->  1m 31s

c:\Users\Harman\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\model_selection\_search.py:320: UserWarning: The total space of parameters 21 is smaller than n_iter=250. Running 21 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 21 candidates, totalling 105 fits


RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),
                   n_iter=250, n_jobs=-1,
                   param_distributions={'bootstrap': [True],
                                        'max_depth': [37, 38, 39],
                                        'max_features': ['sqrt'],
                                        'min_samples_leaf': [1],
                                        'min_samples_split': [5],
                                        'n_estimators': array([260, 270, 280, 290, 300, 310, 320])},
                   random_state=42, return_train_score=True, verbose=2)

### B. Training the model

In [56]:
best_params = random_search.best_params_
print("Best parameters found: ", best_params)

param_grid = {
	"n_estimators": [best_params["n_estimators"] - 5, best_params["n_estimators"], best_params["n_estimators"] + 5],
	"max_depth": [best_params["max_depth"] - 1 if best_params["max_depth"] is not None else None, best_params["max_depth"], best_params["max_depth"] + 1 if best_params["max_depth"] is not None else None],
	"min_samples_split": [max(2, best_params["min_samples_split"] - 1), best_params["min_samples_split"], best_params["min_samples_split"] + 1],
	"min_samples_leaf": [max(1, best_params["min_samples_leaf"] - 1), best_params["min_samples_leaf"], best_params["min_samples_leaf"] + 1],
	"max_features": [best_params["max_features"]],
	"bootstrap": [best_params["bootstrap"]],
}

rf = RandomForestClassifier(random_state=42)

grid_search = GridSearchCV(rf, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2)
grid_search.fit(X_train_balanced, y_train_balanced)

Best parameters found:  {'n_estimators': 280, 'min_samples_split': 5, 'min_samples_leaf': 1, 'max_features': 'sqrt', 'max_depth': 38, 'bootstrap': True}
Fitting 5 folds for each of 81 candidates, totalling 405 fits


GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42), n_jobs=-1,
             param_grid={'bootstrap': [True], 'max_depth': [37, 38, 39],
                         'max_features': ['sqrt'],
                         'min_samples_leaf': [1, 1, 2],
                         'min_samples_split': [4, 5, 6],
                         'n_estimators': [275, 280, 285]},
             verbose=2)

In [57]:
print("Best parameters found by GridSearchCV: ", grid_search.best_params_)

Best parameters found by GridSearchCV:  {'bootstrap': True, 'max_depth': 38, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 5, 'n_estimators': 280}


### G. Evaluating the model

In [58]:
# Balance the test set
# 1. Undersample the majority classes, and keep the values for the minority classes

SAMPLE_TARGET = 1000
undersampling_strategy_test_set = {
    0: SAMPLE_TARGET if y_test.value_counts().get(0) > SAMPLE_TARGET else y_test.value_counts().get(0),	# BENIGN
	1: SAMPLE_TARGET if y_test.value_counts().get(1) > SAMPLE_TARGET else y_test.value_counts().get(1),	# DoS Hulk
	2: SAMPLE_TARGET if y_test.value_counts().get(2) > SAMPLE_TARGET else y_test.value_counts().get(2),	# PortScan
	3: SAMPLE_TARGET if y_test.value_counts().get(3) > SAMPLE_TARGET else y_test.value_counts().get(3),	# DDoS
	4: SAMPLE_TARGET if y_test.value_counts().get(4) > SAMPLE_TARGET else y_test.value_counts().get(4),	# DoS GoldenEye
	5: SAMPLE_TARGET if y_test.value_counts().get(5) > SAMPLE_TARGET else y_test.value_counts().get(5),	# FTP-Patator
	6: SAMPLE_TARGET if y_test.value_counts().get(6) > SAMPLE_TARGET else y_test.value_counts().get(6),	# SSH-Patator
	7: SAMPLE_TARGET if y_test.value_counts().get(7) > SAMPLE_TARGET else y_test.value_counts().get(7),	# DoS slowloris
	8: SAMPLE_TARGET if y_test.value_counts().get(8) > SAMPLE_TARGET else y_test.value_counts().get(8),	# DoS Slowhttptest
	9: SAMPLE_TARGET if y_test.value_counts().get(9) > SAMPLE_TARGET else y_test.value_counts().get(9),	# Bot
	10: SAMPLE_TARGET if y_test.value_counts().get(10) > SAMPLE_TARGET else y_test.value_counts().get(10),	# Web Attack - Brute Force
	11: SAMPLE_TARGET if y_test.value_counts().get(11) > SAMPLE_TARGET else y_test.value_counts().get(11),	# Web Attack - XSS
	12: SAMPLE_TARGET if y_test.value_counts().get(12) > SAMPLE_TARGET else y_test.value_counts().get(12),	# Infiltration
	13: SAMPLE_TARGET if y_test.value_counts().get(13) > SAMPLE_TARGET else y_test.value_counts().get(13),	# Web Attack - SQL Injection
	14: SAMPLE_TARGET if y_test.value_counts().get(14) > SAMPLE_TARGET else y_test.value_counts().get(14),	# Heartbleed
}

rus_test = RandomUnderSampler(random_state=42, sampling_strategy=undersampling_strategy_test_set)
X_test_balanced, y_test_balanced = rus_test.fit_resample(X_test, y_test)


In [59]:
best_rf = grid_search.best_estimator_
y_pred = best_rf.predict(X_test_balanced)

In [60]:
print("Accuracy: ", accuracy_score(y_test_balanced, y_pred))
print("F1 Score: ", f1_score(y_test_balanced, y_pred, average="weighted"))

Accuracy:  0.9763013976098845
F1 Score:  0.9772457305399977


In [61]:
print(classification_report(y_test_balanced, y_pred, target_names=attack_mapping.keys()))

                            precision    recall  f1-score   support

                    BENIGN       0.98      0.97      0.97      1000
                  DoS Hulk       0.99      0.99      0.99      1000
                  PortScan       1.00      1.00      1.00      1000
                      DDoS       1.00      1.00      1.00      1000
             DoS GoldenEye       1.00      1.00      1.00      1000
               FTP-Patator       1.00      1.00      1.00      1000
               SSH-Patator       1.00      1.00      1.00      1000
             DoS slowloris       1.00      0.99      1.00      1000
          DoS Slowhttptest       1.00      0.99      0.99      1000
                       Bot       0.96      1.00      0.98       441
  Web Attack � Brute Force       0.77      0.60      0.68       301
          Web Attack � XSS       0.36      0.57      0.44       119
              Infiltration       0.83      0.71      0.77         7
Web Attack � Sql Injection       0.50      0.75

### **Previous results:**

| Metric             | Previous    | Current     | improvement?                           |
| ------------------ | ----------- | ----------- | -------------------------------------- |
| Accuracy           | 0.95        | 0.98        | <span style="color:#20ff20;">yes</span>|
| MA Range           | 0.95 - 0.95 | 0.89 - 0.91 | <span style="color:#ff4040;">no</span> |
| WA Range           | 0.95 - 0.95 | 0.98 - 0.98 | <span style="color:#20ff20;">yes</span> |
| Precision range    | 0.66 - 1.00 | 0.36 - 1.00 | <span style="color:#ff4040;">no</span> |
| Recall range       | 0.57 - 1.00 | 0.57 - 1.00 | <span style="color:#ffff00;">same</span> |
| F1 range           | 0.63 - 1.00 | 0.44 - 1.00 | <span style="color:#ff4040;">no</span> |
 